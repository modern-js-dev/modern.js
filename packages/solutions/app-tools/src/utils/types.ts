export type DevOptions = {
  entry?: string[] | boolean;
};

export type BuildOptions = {
  analyze?: boolean;
};
