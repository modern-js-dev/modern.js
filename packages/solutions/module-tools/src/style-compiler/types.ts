import { LessOption, SassOption, PostcssOption } from '@modern-js/core';

export type { BuildWatchEmitter } from './build-watch';

export interface Option {
  less?: LessOption;
  sass?: SassOption;
  postcss?: PostcssOption;
}

export interface ResolveItemParams {
  file: string;
  projectDir: string;
  stylesDir: string;
  outDir: string;
  enableVirtualDist: boolean;
  options: Option;
}

export type ResolveParams = Omit<ResolveItemParams, 'file'> & {
  files: string[];
};

export type ProjectOption = Omit<ResolveItemParams, 'file' | 'options'>;

export interface SingleFileCompilerResult {
  code: number; // 0 success, 1 fail
  filename: string;
  content: string;
  error: string | null;
  sourceMapFileName?: string;
  sourceMap?: string;
}

export interface ICompilerResult {
  dists: SingleFileCompilerResult[];
  errors: SingleFileCompilerResult[];
  code: number;
}

export interface IBuildOption {
  projectDir: string;
  stylesDir: string;
  outDir: string;
  enableVirtualDist?: boolean;
  compilerOption?: Option;
}
