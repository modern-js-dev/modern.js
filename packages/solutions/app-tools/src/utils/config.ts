import * as path from 'path';
import { bundle } from '@modern-js/node-bundle-require';
import {
  CONFIG_CACHE_DIR,
  CONFIG_FILE_EXTENSIONS,
  fs,
  getServerConfig,
} from '@modern-js/utils';
import type { NormalizedConfig } from '@modern-js/core';

import type { ServerConfig } from '@modern-js/prod-server';

export const defineServerConfig = (config: ServerConfig): ServerConfig =>
  config;

export const buildServerConfig = async (
  appDirectory: string,
  configFile: string,
  options?: Parameters<typeof bundle>[1],
) => {
  const configFilePath = await getServerConfig(appDirectory, configFile);

  const getOutputFile = (filepath: string) =>
    path.resolve(
      CONFIG_CACHE_DIR,
      `${filepath.replace(
        new RegExp(CONFIG_FILE_EXTENSIONS.join('|')),
        '',
      )}.js`,
    );

  if (configFilePath) {
    await bundle(configFilePath, {
      ...options,
      getOutputFile,
    });
  }
};

export const emitResolvedConfig = async (
  appDirectory: string,
  resolvedConfig: NormalizedConfig,
) => {
  const outputPath = path.join(
    appDirectory,
    resolvedConfig?.output?.path || './dist',
    'modern.config.json',
  );
  await fs.writeJSON(outputPath, resolvedConfig, {
    spaces: 2,
  });
};
