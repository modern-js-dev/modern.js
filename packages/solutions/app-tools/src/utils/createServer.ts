import { Server, ModernDevServerOptions } from '@modern-js/server';

let server: Server | null = null;

export const getServer = () => server;

export const closeServer = async () => {
  if (server) {
    await server.close();
    server = null;
  }
};

export const createServer = async (options: ModernDevServerOptions) => {
  if (server) {
    await server.close();
  }

  // eslint-disable-next-line require-atomic-updates
  server = new Server(options);

  const app = await server.init();

  return app;
};
