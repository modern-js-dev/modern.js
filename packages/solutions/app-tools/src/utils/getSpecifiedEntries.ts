import inquirer from 'inquirer';
import { Entrypoint } from '@modern-js/types';

export const getSpecifiedEntries = async (
  entry: string[] | boolean,
  entrypoints: Entrypoint[],
): Promise<string[]> => {
  const entryNames = entrypoints.map(e => e.entryName);
  if (!entry) {
    return entryNames;
  }

  if (typeof entry === 'boolean') {
    const { selected } = await inquirer.prompt([
      {
        type: 'checkbox',
        name: 'selected',
        choices: entryNames,
        message: '请选择需要构建的入口',
        validate(answer: string[]) {
          if (answer.length < 1) {
            return 'You must choose at least one topping.';
          }

          return true;
        },
      },
    ]);

    return selected;
  }

  entry.forEach(name => {
    if (!entryNames.includes(name)) {
      throw new Error(
        `can not found entry ${name}, compiler entry should in ${entryNames.join(
          ', ',
        )}`,
      );
    }
  });

  return entry;
};
