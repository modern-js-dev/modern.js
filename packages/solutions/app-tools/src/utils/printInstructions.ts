import { prettyInstructions, logger, isDev, chalk } from '@modern-js/utils';
import type {
  IAppContext,
  NormalizedConfig,
  ToRunners,
  CliHooks,
} from '@modern-js/core';

export const printInstructions = async (
  hookRunners: ToRunners<CliHooks>,
  appContext: IAppContext,
  config: NormalizedConfig,
) => {
  let message = prettyInstructions(appContext, config);
  const { existSrc } = appContext;

  if (isDev() && existSrc) {
    message += `\n${chalk.cyanBright(
      [
        `Note that the development build is not optimized.`,
        `To create a production build, execute build command.`,
      ].join('\n'),
    )}`;
  }

  // call beforePrintInstructions hook.
  const { instructions } = await hookRunners.beforePrintInstructions({
    instructions: message,
  });

  logger.log(instructions);
};
