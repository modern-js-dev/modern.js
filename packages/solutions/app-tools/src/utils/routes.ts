import path from 'path';
import { IAppContext } from '@modern-js/core';
import { fs, ROUTE_SPEC_FILE } from '@modern-js/utils';

const generateRoutes = async (appContext: IAppContext) => {
  const { serverRoutes, distDirectory } = appContext;
  const output = JSON.stringify({ routes: serverRoutes }, null, 2);
  await fs.writeFile(path.join(distDirectory, ROUTE_SPEC_FILE), output);
};

export { generateRoutes };
