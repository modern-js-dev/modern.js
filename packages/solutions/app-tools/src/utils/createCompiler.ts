import webpack, { Configuration, StatsCompilation } from 'webpack';
import { IAppContext, NormalizedConfig, mountHook } from '@modern-js/core';
import {
  chalk,
  logger,
  formatWebpackMessages,
  clearConsole,
} from '@modern-js/utils';
import { printInstructions } from './printInstructions';

const prettyTime = (stats: StatsCompilation) =>
  Math.max(...(stats.children?.map(child => child.time || 0) || []));

export const createCompiler = async ({
  webpackConfigs,
  // TODO: params
  userConfig,
  appContext,
}: {
  webpackConfigs: Configuration[];
  userConfig: NormalizedConfig;
  appContext: IAppContext;
}) => {
  try {
    await (mountHook() as any).beforeCreateCompiler({ webpackConfigs });
    const compiler = webpack(webpackConfigs);

    await (mountHook() as any).afterCreateCompiler({ compiler });

    let isFirstCompile = true;

    compiler.hooks.invalid.tap('invalid', () => {
      clearConsole();
      logger.log('Compiling...');
    });

    compiler.hooks.done.tap('done', async (stats: any) => {
      clearConsole();

      const statsData = stats.toJson({
        all: false,
        warnings: true,
        errors: true,
        timings: true,
      });

      const { errors, warnings } = formatWebpackMessages(statsData);

      if (errors.length) {
        logger.log(chalk.red(`Failed to compile.\n`));
        logger.log(errors.join('\n\n'));
        logger.log();
      } else if (process.stdout.isTTY || isFirstCompile) {
        await (mountHook() as any).afterDev();
        if (warnings.length) {
          logger.log(chalk.yellow(`Compiled with warnings.\n`));
          logger.log(warnings.join('\n\n'));
          logger.log();
        } else {
          logger.log(
            chalk.green(
              `Compiled successfully in ${prettyTime(statsData)} ms.\n`,
            ),
          );
        }
        await printInstructions(appContext, userConfig);
      }
      // eslint-disable-next-line require-atomic-updates
      isFirstCompile = false;
    });

    return compiler;
  } catch (err) {
    logger.log(chalk.red(`Failed to compile.`));
    logger.log();
    logger.log(err as any);
    // eslint-disable-next-line no-process-exit
    process.exit(1);
  }
};
