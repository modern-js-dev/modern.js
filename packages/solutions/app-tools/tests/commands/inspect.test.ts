import { fs } from '@modern-js/utils';
import { NormalizedConfig } from '@modern-js/core';
import { IAppContext } from '@modern-js/types';
import {
  webpack,
  Configuration,
  WebpackConfigTarget,
} from '@modern-js/webpack';
import {
  printInspectResult,
  formatWebpackConfig,
  getTagByWebpackTarget,
} from '../../src/commands/inspect';

describe('inspect command', () => {
  const { DefinePlugin } = webpack;

  const webpackConfig: Configuration = {
    mode: 'development',
    plugins: [new DefinePlugin({ foo: 'bar' })],
  };

  const appContext = {
    metaName: '',
    appDirectory: __dirname,
    distDirectory: `${__dirname}/dist`,
    srcDirectory: `${__dirname}/src`,
    sharedDirectory: `${__dirname}/src/shared`,
    internalSrcAlias: '@_modern_js_src',
    internalDirAlias: '@_modern_js_internal',
    internalDirectory: `${__dirname}/node_modules/.modern-js`,
  } as IAppContext;

  const normalizedConfig = {
    source: {
      configDir: '/config',
    },
    output: {
      path: '/',
      jsPath: '/js',
      cssPath: '/css',
    },
  } as NormalizedConfig;

  test('should format webpack config correctly', () => {
    expect(formatWebpackConfig(webpackConfig)).toMatchSnapshot();
  });

  test('should get correct tag by webpack target', () => {
    expect(getTagByWebpackTarget(WebpackConfigTarget.CLIENT)).toEqual('client');
    expect(getTagByWebpackTarget(WebpackConfigTarget.NODE)).toEqual('ssr');
    expect(getTagByWebpackTarget(WebpackConfigTarget.MODERN)).toEqual('modern');
    expect(() => {
      getTagByWebpackTarget('foo' as any);
    }).toThrowError();
  });

  test('should log result in console', () => {
    const log = jest.fn();
    const logSpy = jest.spyOn(console, 'log').mockImplementation(log);
    const fsSpy = jest.spyOn(fs, 'outputFileSync');

    printInspectResult(
      WebpackConfigTarget.CLIENT,
      appContext,
      normalizedConfig,
      {},
    );

    expect(log).toHaveBeenCalledTimes(1);
    logSpy.mockRestore();
    fsSpy.mockRestore();
  });

  test('should not log result in console when console option is false', () => {
    const log = jest.fn();
    const logSpy = jest.spyOn(console, 'log').mockImplementation(log);
    const fsSpy = jest.spyOn(fs, 'outputFileSync');

    printInspectResult(
      WebpackConfigTarget.CLIENT,
      appContext,
      normalizedConfig,
      {
        console: false,
      },
    );

    expect(log).toHaveBeenCalledTimes(0);
    logSpy.mockRestore();
    fsSpy.mockRestore();
  });
});
