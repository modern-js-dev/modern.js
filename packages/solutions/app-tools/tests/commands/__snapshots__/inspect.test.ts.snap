// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inspect command should format webpack config correctly 1`] = `
"module.exports = {
  mode: 'development',
  plugins: [
    {
      definitions: {
        foo: 'bar'
      }
    }
  ]
};"
`;
